ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto

```diff
@@ -202,3 +206,33 @@ message CreateAndExerciseCommand {
   // Required
   Value choice_argument = 4;
 }
+
+// An additional contract that is used to resolve
+// contract & contract key lookups.
+message DisclosedContract {
```

Comment:
@cocreature This is missing the contract key value. Is that an oversight, or do we expect the ledger api server to compute the contract key value from the arguments? If yes, can you please point me to code that would compute the key?

Reply:
We deliberately did not include the key value since it seemed like a fair amount of data to ship around. I don’t actually see why you need to compute it. My assumption was that the engine queries the ledger for a key via ResultNeedKey as it does currently. You then search by key hash not key value. To validate that the key hash is correct, you set a flag on the result you pass to the engine and the engine will then recompute the key & validate.

What do you need the key value for (before calling daml engine, afterwards it’s in the transaction)?

Reply:
Oh I see. I was confused because we have half a dozen internal calls between `ResultNeedKey` and the actual database lookup. All the intermediate layers take a `com.daml.lf.transaction.GlobalKey` which includes the key value, but we only actually need the key hash. I'll look into refactoring that.

Reply:
> Oh I see. I was confused because we have half a dozen internal calls between ResultNeedKey and the actual database lookup.

Depending on how you wire it up, you don’t necessarily need to change the callpath to the database: The engine gives you both the key & hash on ResultNeedKey. You can use the key hash to resolve against the disclosed contracts but continue passing down the value to the database if that lookup fails. Not saying refactoring is a bad idea though, I’ll leave that up to you.